working/0546-patterns/goals-and-constraints.md

```diff
@@ -0,0 +1,461 @@
+use that type variable in things like generic constructor calls.
+
+[type pattern]: https://github.com/dart-lang/language/issues/170
+
+### Guard clauses
+
+This isn't *quite* tied to patterns, but most languages with a pattern matching
+construct also support *guard clauses*. These are arbitrary predicate
+expressions you place after the pattern. A pattern only matches if that
+expression evaluates to true.
+
+In C#, these are [`when` clauses][csharp when]. In Haskell, [they appear after
+a `|`][haskell guard].
+
+[csharp when]: https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching#when-clauses-in-case-expressions
+
+[haskell guard]: http://learnyouahaskell.com/syntax-in-functions
+
+## Uses for Patterns
+
+Once you have patterns, they need to be embedded in some other language
+constructs that developers can actually use. I've mentioned a few, but I'll run
+quickly through places we should or could consider:
+
+*   **A pattern matching statement.** Like a switch statement but where each
+    case clause is a pattern instead of a simple constant expression.
+
+*   **A pattern matching *expression*.** Statements are fine for imperative
+    code, but people using patterns often like to write in a functional,
+    expression-oriented style. In that case, it might be good to have a pattern
+    matching construct that can be embedded in an expression.
+
+    Match expressions rely on good exhaustiveness checking. A statement can
+    simply do nothing if no case matches. An expression must evaluate to *some*
+    value (or throw an exception, which is rarely helpful). That means ensuring
+    that at least one case matches and telling the user at compile time if that
+    may not happen.
+
+*   **A declaration statement.** If you have some object that you know has a
+    certain structure and you just want to destructure it and pull out some data
+    bound to new variables, it's very handy to have a variable declaration
+    statement form that takes a single pattern and matches it against the
+    initializer.
+
+    Most languages with pattern matching use this for all variable declarations.
+
+*   **For-in loop variable clauses.** A for-in loop is another place where a
+    variable can be declared. We could extend that to support arbitrary patterns.
+    This might be particularly useful for iterating over the entries in a map:
+
+    ```dart
+    for (var (key, value) in someMap) {
+      ...
+    }
+    ```
+
+*   **A pattern-based if.** This is basically a lightweight, single case pattern
+    match statement. You have one pattern and a "then branch" you execute if the
+    pattern matches and an optional else branch if it doesn't.
+
+    For example, [C# allows a type pattern][csharp] inside the condition of an
+    if statement and Swift has [`if case`][swift].
+
+    [csharp]: https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching
+    [swift]: http://fuckingifcaseletsyntax.com/
```

Comment:
😮 Had to parse that URL 3 times...

Reply:
Yeah, a little unfortunate that this was the best reference I could find:

*Achievement unlocked: Used a profane URL in a language proposal.*